lib/vrl/compiler/src/type_def.rs

```diff
@@ -349,8 +349,22 @@ impl TypeDef {
     }
 
     pub fn merge(&mut self, other: Self, strategy: merge::Strategy) {
-        self.kind.merge(other.kind, strategy);
         self.fallible |= other.fallible;
+
+        // NOTE: technically we shouldn't do this, but to keep backward compatibility with the
+        // "old" `Kind` implementation, we consider a type that is marked as "any" equal to the old
+        // implementation's `unknown`, which, when merging, would discard that type.
+        //
+        // see: https://github.com/vectordotdev/vector/blob/18415050b60b08197e8135b7659390256995e844/lib/vrl/compiler/src/type_def.rs#L428-L429
```

Comment:
It's not clear to me which "right" way you envision here or does this comment not imply that we should change any of this logic in the future?

If we intend to change it, I'd suggest opening an issue with the proposed logic and a migration plan.

Reply:
I'll likely introduce a new merge `Strategy` option that allows ignoring `any` values. It's not "bad" per sé, but it is confusing behaviour without explicitly requesting it as the caller of a type definition merge.

For now, that strategy doesn't exist in the `value` crate, so it lives here in VRL, I'll think on what the best next steps would be to either remove the need for this entirely, or move it to a more "official" place as a strategy in the value crate.